### ISSUE TITLE
FSVINN retraining crashes at validation/testing stage with allocation memory error, but not during training

### ISSUE THREAD

**Issue by mafortin** [Deep-MI/FastSurfer#364]
## Question/Support Request

As the title says, I am retraining FastSurfer-VINN 2D U-Net with some (0.7mm)3 HCP data and the training is going smoothly with a batch size of 4 for (256)3 matrix size (n_train=75), but I get 'Can't allocate memory' error at the evaluation stage of the training (at the 1st epoch). It actually happens at the calculation of the MIoUs in the train.py script. I use a batch size of 1 for testing for (256)3 matrix sizes (n_val=10). See screenshot below for more details.

I've been reading online about this issue of allocation of memory at the validation stage only and I tried to add the 'with torch.no_grad() before the evaluation stage but it didn't change the error. I also saw at the same time that the 'flag' @torch.no_grad() was already there in the script.

Any idea/suggestion? I assumed I could not go lower than a batch size of 1 for validation also.

## Screenshots

![image](https://github.com/Deep-MI/FastSurfer/assets/51452174/82737ab2-3de4-4155-be14-3af0ac0301b8)

## Environment
 - FastSurfer Version: v2.0.4 (i think) developer version (I cloned the dev branch on 04/05/23).
 - OS: Linux Ubuntu 20.04
 - CPU: 64 Gb - 16 cores
 - GPU: Tesla P100-PCie 16 Gb

pip freeze:
![image](https://github.com/Deep-MI/FastSurfer/assets/51452174/5f89b42c-3e48-41b9-9f1f-757bdbb12a21)

PyCharm config to run the script run_model.py:
![image](https://github.com/Deep-MI/FastSurfer/assets/51452174/d05b8737-71a5-4779-90eb-536b18024741)
![image](https://github.com/Deep-MI/FastSurfer/assets/51452174/1317b90e-ec85-4e00-a2b8-e585d6686568)

**Comment by LeHenschel** [comment]
The validation uses the same batch size setting as the training. Testing is for inference only. See here: https://github.com/Deep-MI/FastSurfer/blob/stable/FastSurferCNN/data_loader/loader.py#L162. You can try to set this to the Test.batch_size for the validation.

Would be helpful to know how much memory is already allocated by the network, the training phase, and finally the validation stage on your GPUs.

**Comment by dkuegler** [comment]
I am not sure, how you modify your data loader, but the script seems to crash in the data loader.

Also you can start your training in pycharm debug mode and print `torch.cuda.memory_summary() ` to the console or even just insert this line to debug.

Alternatively, you may need to look into the memory profiler https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html

**Comment by mafortin** [comment]
Hi @LeHenschel and @dkuegler ! :)

I originally thought it was a GPU memory issue, but I arrived to the conclusion it was a CPU allocation issue. After clearing some background processes, I've been able to run it, and realized that there is a considerable peak of RAM usage at the beginning of the data_loader usage in the validation stage. 

Thus, I do not think it's a FastSurfer issue per se, but a general RAM allocation memory issue from my end. 

Sorry for the inconvenience! :)
